roaringbitmap/src/main/java/org/roaringbitmap/SparseOrderedWriter.java

```diff
@@ -0,0 +1,175 @@
+ * The values are expected to be (increasing) sorted order.
+ * Values are first written to a temporary internal buffer, but
+ * the underlying bitmap can forcefully synchronize by calling "flush"
+ * (although calling flush to often would defeat the performance
+ * purpose of this class).
+ * The main use case for an SparseOrderedWriter is to create bitmaps quickly.
+ * You should benchmark your particular use case to see if it helps.
+ *
+ * <pre>
+ * {@code
+ *
+ *     //...
+ *
+ *     RoaringBitmap r = new RoaringBitmap();
+ *     SparseOrderedWriter ow = new SparseOrderedWriter(r);
+ *     for (int i :....) {
+ *     ow.add(i);
+ *     }
+ *     ow.flush(); // important
+ * }
+ * </pre>
+ */
+public class SparseOrderedWriter implements OrderedWriter {
+
+  private static final int ARRAY_MAX_SIZE = ArrayContainer.DEFAULT_MAX_SIZE;
+
+  private final RoaringBitmap underlying;
+  private short[] values;
+  private int valuesCount;
+  private short currentKey;
+
+  /**
+   * Initialize an SparseOrderedWriter with a receiving bitmap
+   *
+   * @param underlying bitmap where the data gets written
+   */
+  public SparseOrderedWriter(RoaringBitmap underlying) {
+    this(underlying, ARRAY_MAX_SIZE);
+  }
+
+  /**
+   * Initialize an SparseOrderedWriter with a receiving bitmap
+   *
+   * @param underlying bitmap where the data gets written
+   * @param size       maximum amount of values per container
+   */
+  SparseOrderedWriter(RoaringBitmap underlying, int size) {
+    this.underlying = underlying;
+    this.valuesCount = 0;
+    this.values = new short[size];
+  }
+
+  /**
+   * Initialize an SparseOrderedWriter and construct a new RoaringBitmap
+   */
+  public SparseOrderedWriter() {
+    this(new RoaringBitmap());
+  }
+
+  /**
+   * Grab a reference to the underlying bitmap
+   *
+   * @return the underlying bitmap
+   */
+  public RoaringBitmap getUnderlying() {
+    return underlying;
+  }
+
+  /**
+   * Adds the value to the underlying bitmap. The data might
+   * be added to a temporary buffer. You should call "flush"
+   * when you are done.
+   *
+   * @param value the value to add.
+   * @throws IllegalStateException if values are not added in increasing order
+   */
+  @Override
+  public void add(int value) {
+    short key = Util.highbits(value);
+    short low = Util.lowbits(value);
+    if (key != currentKey) {
+      if (Util.compareUnsigned(key, currentKey) < 0) {
+        throw new IllegalStateException("Must write in ascending key order");
+      }
+      flush();
+    }
+
+    int valueIndex = valuesCount++;
+    if (valueIndex > 0 && Util.compareUnsigned(low, values[valueIndex - 1]) <= 0) {
+      throw new IllegalStateException("Must write in ascending order.");
+    }
+    if (valueIndex >= values.length) {
+      valuesCount = values.length;
+      throw new IllegalStateException(
+          String.format("Exceeded limit of %d values per container.", values.length));
+    }
+
+    currentKey = key;
```

Comment:
move this into the block where the flush occurs to make it more obvious when this value can change